src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -39,90 +39,270 @@
+        this.archivedRecordingsReportPath = webServerTempPath;
         this.targetConnectionManager = targetConnectionManager;
         this.clock = clock;
         this.platformClient = platformClient;
-        this.reportService = reportService;
+        this.notificationFactory = notificationFactory;
+        this.base32 = base32;
     }
 
-    public String saveRecording(ConnectionDescriptor connectionDescriptor, String recordingName)
-            throws Exception {
+    public Future<String> saveRecording(
+            ConnectionDescriptor connectionDescriptor, String recordingName) {
 
-        String saveName =
-                targetConnectionManager.executeConnectedTask(
-                        connectionDescriptor,
-                        connection -> {
-                            Optional<IRecordingDescriptor> descriptor =
-                                    this.getDescriptorByName(connection, recordingName);
+        CompletableFuture<String> future = new CompletableFuture<>();
 
-                            if (descriptor.isPresent()) {
-                                return writeRecordingToDestination(connection, descriptor.get());
-                            } else {
-                                throw new RecordingNotFoundException(recordingName);
-                            }
-                        });
+        try {
+            String saveName =
+                    targetConnectionManager.executeConnectedTask(
+                            connectionDescriptor,
+                            connection -> {
+                                Optional<IRecordingDescriptor> descriptor =
+                                        this.getDescriptorByName(connection, recordingName);
 
-        return saveName;
+                                if (descriptor.isPresent()) {
+                                    return writeRecordingToDestination(
+                                            connection, descriptor.get());
+                                } else {
+                                    throw new RecordingNotFoundException(
+                                            "active recordings", recordingName);
+                                }
+                            });
+            future.complete(saveName);
+            notificationFactory
+                    .createBuilder()
+                    .metaCategory(SAVE_NOTIFICATION_CATEGORY)
+                    .metaType(HttpMimeType.JSON)
+                    .message(
+                            Map.of(
+                                    "recording",
+                                    saveName,
+                                    "target",
+                                    connectionDescriptor.getTargetId()))
+                    .build()
+                    .send();
+        } catch (RecordingNotFoundException
+                | IOException
+                | URISyntaxException
+                | FlightRecorderException e) {
```

Comment:
Can this `catch` be merged with the one below?

Reply:
I was thinking to keep the specific `Exceptions` separate from the generic `Exception` because they'll be wrapped inside an `ExecutionException` but I guess that doesn't matter since the caller can just use `ExceptionUtils.getRootCause(e)` to determine the specific `Exception` thrown.

Reply:
In both branches the Future is completed exceptionally, so the caller would get an `ExecutionException` on the `.get()` call in either case. So I think collapsing these two clauses together actually won't change anything for callers.